qiskit/primitives/containers/bit_array.py

```diff
@@ -347,3 +368,200 @@ def reshape(self, *shape: ShapeInput) -> "BitArray":
         else:
             raise ValueError("Cannot change the size of the array.")
         return BitArray(self._array.reshape(shape), self.num_bits)
+
+    def transpose(self, axes: tuple[int, ...] | list[int] | None = None) -> "BitArray":
+        """Return a bit array with axes transposed.
+
+        Args:
+            axes: tuple or list of ints, optional. See
+                `numpy.transpose
+                <https://numpy.org/doc/stable/reference/generated/numpy.transpose.html>`_
+                for the details.
+
+        Returns:
+            BitArray: a bit array with axes permuted.
+
+        Raises:
+            ValueError: if ``axes`` don't match this bit array.
+            ValueError: if ``axes`` includes any indices that are out of bounds.
+        """
+        if axes is None:
+            axes = tuple(reversed(range(self.ndim)))
+        if len(axes) != self.ndim:
+            raise ValueError("axes don't match bit array")
+        for i in axes:
+            if i >= self.ndim or self.ndim + i < 0:
+                raise ValueError(
+                    f"axis {i} is out of bounds for bit array of dimension {self.ndim}."
+                )
+        axes = tuple(i if i >= 0 else self.ndim + i for i in axes) + (-2, -1)
+        return BitArray(self._array.transpose(axes), self.num_bits)
+
+    def marginalize(self, indices: int | Sequence[int]) -> "BitArray":
+        """Return a bit array marginalized over some indices of interests.
+
+        Args:
+            indices: The bit positions of interest to marginalize over.
+
+        Returns:
+            A bit array marginalized to bits of interests.
+
+        Raises:
+            ValueError: if there are any invalid indices to marginalize over.
+        """
+        if isinstance(indices, int):
+            indices = (indices,)
+        for index in indices:
+            if index < 0 or index >= self.num_bits:
+                raise ValueError(
+                    f"index {index} is out of bounds for the number of bits {self.num_bits}."
+                )
+        arr = _unpack(self)
+        arr = arr[..., indices]
+        arr, num_bits = _pack(arr)
+        return BitArray(arr, num_bits)
+
+    def expectation_value(
```

Comment:
In the meeting we had discussed an array-valued version of this method. I'm not sure if it's possible to introduce such a thing in this PR because it would be blocked by making ObservableArray a public class. @chriseclectic may have advice.

Reply:
Since `sample_expval_{float,complex}` returns a single expval, we need to extend it to allow multiple operators and return multiple expvals. It looks technically easy, but it might be good to be a separate PR.

As the first step, it might be nice to extend this `expectation_value` method to handle `ObservableArray` itself or a list generated by `ObservableArray.tolist()` without changing the rust code, i.e., run a loop in `expectation_value` method.

As the second step, we can update the rust code optimized for our input of `expectaion_value` method and remove the loop in `expectation_value` method.

https://github.com/Qiskit/qiskit/blob/5fb343f5f2faf5578dc0afcd6a3393fbaa0e0c17/crates/accelerate/src/sampled_exp_val.rs#L58-L70

Reply:
One way to interpret the current implementation is that it's implementing the trivial broadcast rule. Ie, the "shape" of the input observable is `()`. We could look to extend this method by, in a future PR, accepting a non-trivial shape of observables, like an ObservablesArray or something coercible thereto.

With this in mind, what do you think about removing the `loc` argument, and instead returning the output of `sampled_expectation_value` for _every_ index of the shape. For example, if a bit array has shape (2,5), then the output of this method will be a float array of shape (2,5)? The extension to multiple observables would be to demand broadcastability between the observables array and the bit array.

Reply:
Retuning an array of expval with the same shape makes sense to me. If users want to compute expvals of a subset of an array, they can slice off a part of their bit array using the new methods introduced in this PR.
Which idea do you have in mind, returning an array of expvals or a data bin of evs and stds as same as Estimator V2 output?

Reply:
I would say just the array. The stds could be derived from the array in the usual way (`1-ev^2` divided by shots etc etc), if the user requires it. Perhaps this could be mentioned in the docstring?


```python
def expectation_values(self, operator: ObservablesLike) -> NDArray[np.float64]:
    """Compute the expectation value the operator for every entry of this bit array.

    Args:
         operator: The operator to take the expectation value of. 
    
    Returns:
        An array of expectation values whose shape matches this bit array's :attr:`~shape`.
    """
```

with the goal of moving to

```python
def expectation_values(self, operator: ObservablesArrayLike) -> NDArray[np.float64]:
    """Compute the expectation values of the provided observables, broadcasted against this bit array.

    Args:
         operator: The operator(s) to take the expectation value of. Must have a shape broadcastable with
            with this bit array.
    
    Returns:
        An array of expectation values whose shape is the broadcast shape of ``operator`` and this bit array.
        
    Raises:
        ValueError: If the provided operator does not have a shape broadcastable with this bit array.
    """
```

I wonder if we could get this second version into 1.1? Probably not enough time.

Reply:
I implemented the observable array version in the latest commit. Could you take a look at it?